File: algo/53. Maximum Subarray.js
// @9 Array

/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    var sum = 0;
    var minSum = 0;
    var res = -Infinity;
    for (var i = 0; i < nums.length; i++) {
        sum += nums[i];
        if (sum - minSum > res) {
            res = sum - minSum;
        }
        if (sum < minSum) {
            minSum = sum;
        }
    }
    return res;
};

File: round2/53. Maximum Subarray.cs
public class Solution {
    public int MaxSubArray(int[] nums) {
        if (nums.Length == 0) return 0;
        int currSum = nums[0];
        int res = currSum;
        for (int i=1; i<nums.Length; i++) {
            currSum = currSum + nums[i] > nums[i] ? currSum + nums[i] : nums[i];
            res = res > currSum ? res : currSum;
        }
        return res;
    }
}